tree_assume/src/optimizations/is_valid.rs

```diff
@@ -0,0 +1,65 @@
+use crate::schema_helpers::{Constructor, ESort, Purpose};
+use strum::IntoEnumIterator;
+
+fn rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let actions = ctor.filter_map_fields(|field| match field.purpose {
+        Purpose::Static(_) => None,
+        Purpose::CapturedExpr | Purpose::SubExpr | Purpose::SubListExpr => Some(format!(
+            "({sort}IsValid {var})",
+            sort = field.sort().name(),
+            var = field.var()
+        )),
+    });
+
+    if actions.is_empty() {
+        return None;
+    }
+    let actions = actions.join("\n");
+    let pat = ctor.construct(|field| field.var());
+    let sort = ctor.sort().name();
+    Some(format!(
+        "(rule (({sort}IsValid {pat})) ({actions}) :ruleset always-run)"
+    ))
+}
+
+pub(crate) fn rules() -> Vec<String> {
+    ESort::iter()
+        .map(|sort| "(relation *IsValid (*))".replace('*', sort.name()))
+        .chain(Constructor::iter().filter_map(rule_for_ctor))
+        .collect::<Vec<_>>()
+}
+#[cfg(test)]
+use crate::ast::*;
+#[cfg(test)]
+use crate::Value;
+
```

Comment:
Might be good to have a test that shows that validity passes through the assume node but not its left child

Reply:
Done